lib/plugins/threaded-signing/plugin.c

```diff
@@ -137,29 +152,61 @@ signing_worker_thread(void *user_data)
   g_cond_signal(&self->cond);
 
   while (self->is_running) {
-    // Wait for a signal, triggered when it is time to sign a hash.
-    g_cond_wait(&self->cond, &self->mutex);
-
-    if (self->plugin_state == THREADED_SIGNING_HAS_HASH_TO_SIGN) {
-      // Copy the |hash_to_sign| to |signature_info| and start signing. In principle, it is now
-      // possible to prepare for a new |hash_to_sign|.
+    if (self->input_buffer_idx > 0) {
+      // Get the oldest hash from the input buffer
+      // Copy the hash to |signature_info| and start signing. In principle, it is now possible to
+      // prepare for a new hash.
       assert(self->hash_size == self->signature_info->hash_size);
-      memcpy(self->signature_info->hash, self->hash_to_sign, self->hash_size);
-      self->plugin_state = THREADED_SIGNING_SIGNS_HASH;
+      assert(self->signature_info->hash);
+      memcpy(self->signature_info->hash, self->input_buffer[0], self->hash_size);
+
+      uint8_t *tmp = self->input_buffer[0];
+      int j = 0;
+      while (self->input_buffer[j] != NULL && j < MAX_BUFFER_LENGTH) {
```

Comment:
This will still make you read outside memory if the entire buffer is filled with pointers to allocated memories.
You need to stop one element earlier.